### PR TITLE
Disable Jackson default JSON processing limits

### PR DESCRIPTION
**Pull request by wendigo**
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

Fixes https://github.com/trinodb/trino/issues/17843

### CODE REVIEW

**Comment by wendigo** [comment]
I tried to create smallest change possible

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/JsonTypeUtil.java [hunk 1]
```diff
@@ -37,8 +34,9 @@
 
 public final class JsonTypeUtil
 {
-    private static final JsonFactory JSON_FACTORY = new JsonFactoryBuilder().disable(CANONICALIZE_FIELD_NAMES).build();
-    private static final ObjectMapper SORTED_MAPPER = new ObjectMapperProvider().get().configure(ORDER_MAP_ENTRIES_BY_KEYS, true);
+    private static final ObjectMapper SORTED_MAPPER = new ObjectMapperProvider()
```
Comment (electrum) [inline lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/JsonTypeUtil.java:37]: Unrelated formatting change

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/JsonUtils.java [hunk 2]
```diff
@@ -36,7 +36,8 @@
 
 public final class JsonUtils
 {
-    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapperProvider().get()
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapperProvider()
```
Comment (electrum) [inline lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/JsonUtils.java:39]: Unrelated formatting change

**Review by electrum — APPROVED** [review]

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/DefaultJsonFactory.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ */
+package io.trino.plugin.base.util;
+
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonFactoryBuilder;
+import io.airlift.json.ObjectMapperProvider;
+
+import static com.google.common.base.Verify.verify;
+
+public class DefaultJsonFactory
```
Comment (findepi) [inline lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/DefaultJsonFactory.java:22]: utility classes should be final

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/DefaultJsonFactory.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+{
+    public static final JsonFactory INSTANCE = new ObjectMapperProvider()
+            .get()
+            .getFactory();
```
Comment (findepi) [inline lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/DefaultJsonFactory.java:26]: `ObjectMapperProvider` is focused on object mapping and configures that.
the fact that it disables string limits looks incidental from API perspective.
also, `ObjectMapperProvider` used for REST APIs probably should have some sane limits restored.

for internal query processing (eg json functions) we don't want limits. thus i think we should instantiate and configure `JsonFactory` directly, without calling upon `ObjectMapperProvider`

File: service/trino-verifier/src/main/java/io/trino/verifier/JsonEventClient.java [hunk 5]
```diff
@@ -29,6 +30,7 @@
 import static java.nio.charset.Charset.defaultCharset;
 import static java.util.Objects.requireNonNull;
 
+@SuppressModernizer // Serialization path is not affected by Jackson 2.15 constraints
```
Comment (findepi) [inline service/trino-verifier/src/main/java/io/trino/verifier/JsonEventClient.java:33]: narrow scope of suppression

File: testing/trino-benchmark/src/main/java/io/trino/benchmark/JsonBenchmarkResultWriter.java [hunk 6]
```diff
@@ -33,7 +33,7 @@ public JsonBenchmarkResultWriter(OutputStream outputStream)
     {
         requireNonNull(outputStream, "outputStream is null");
         try {
-            jsonGenerator = new JsonFactory().createGenerator(outputStream, JsonEncoding.UTF8);
+            jsonGenerator = DefaultJsonFactory.INSTANCE.createGenerator(outputStream, JsonEncoding.UTF8);
```
Comment (hashhar) [inline testing/trino-benchmark/src/main/java/io/trino/benchmark/JsonBenchmarkResultWriter.java:36]: isn't this also serialization path?

File: testing/trino-benchmark/src/main/java/io/trino/benchmark/OdsBenchmarkResultWriter.java [hunk 7]
```diff
@@ -36,7 +36,7 @@ public OdsBenchmarkResultWriter(String entity, OutputStream outputStream)
         requireNonNull(outputStream, "outputStream is null");
         this.entity = entity;
         try {
-            jsonGenerator = new JsonFactory().createGenerator(outputStream, JsonEncoding.UTF8);
+            jsonGenerator = DefaultJsonFactory.INSTANCE.createGenerator(outputStream, JsonEncoding.UTF8);
```
Comment (hashhar) [inline testing/trino-benchmark/src/main/java/io/trino/benchmark/OdsBenchmarkResultWriter.java:39]: isn't this serialization path?

**Review by hashhar — APPROVED** [review]
Looks good % comments.

re: allowing default factory on serialization paths - we won't know when code is introduced in the serialization places which also does de-serialization. Why not use the no-limits JsonFactory in serialization path as well - doesn't seem to hurt?

File: client/trino-cli/src/main/java/io/trino/cli/JsonPrinter.java [hunk 8]
```diff
@@ -70,4 +72,17 @@ private static Object formatValue(Object o)
         }
         return o;
     }
+
+    @SuppressModernizer
```
Comment (kokosing) [inline client/trino-cli/src/main/java/io/trino/cli/JsonPrinter.java:76]: Please provide a comment why it is suppressed

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/DefaultJsonFactory.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+ * limitations under the License.
+ */
+package io.trino.plugin.base.util;
+
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonFactoryBuilder;
+import com.fasterxml.jackson.core.StreamReadConstraints;
+import org.gaul.modernizer_maven_annotations.SuppressModernizer;
+
+public final class DefaultJsonFactory
```
Comment (kokosing) [inline lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/DefaultJsonFactory.java:21]: TrinoJsonFactory?

File: lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/json/JsonSerializer.java [hunk 10]
```diff
@@ -67,7 +68,7 @@ public JsonSerializer(List<Column> columns)
                 .map(column -> field(column.name().toLowerCase(Locale.ROOT), column.type()))
                 .collect(toImmutableList()));
 
-        jsonFactory = new JsonFactory();
+        jsonFactory = DefaultJsonFactory.INSTANCE;
```
Comment (kokosing) [inline lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/json/JsonSerializer.java:71]: You can remove the `jsonFactory` field

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/DefaultJsonFactory.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+{
+    public static final JsonFactory INSTANCE = builder()
```
Comment (kokosing) [inline lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/DefaultJsonFactory.java:23]: Maybe you can rename `INSTANCE` to `JSON_FACTORY` and then you can use static imports

File: core/trino-main/src/main/java/io/trino/operator/scalar/JsonExtract.java [hunk 12]
```diff
@@ -118,7 +118,7 @@
 {
     private static final int ESTIMATED_JSON_OUTPUT_SIZE = 512;
 
-    private static final JsonFactory JSON_FACTORY = new JsonFactoryBuilder()
+    private static final JsonFactory JSON_FACTORY = DefaultJsonFactory.builder()
             .disable(CANONICALIZE_FIELD_NAMES)
             .build();
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/operator/scalar/JsonExtract.java:123]: maybe we can have a method or single json factory field that is shared between all json functions (all 4 java files)?

Or maybe we should define another field in `DefaultJsonFactory` that we can simply use here?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionData.java [hunk 13]
```diff
@@ -37,7 +38,7 @@
         implements StructLike
 {
     private static final String PARTITION_VALUES_FIELD = "partitionValues";
-    private static final JsonFactory FACTORY = new JsonFactory();
+    private static final JsonFactory FACTORY = DefaultJsonFactory.INSTANCE;
```
Comment (kokosing) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionData.java:41]: inline field

**Review by kokosing — COMMENTED** [review]
This change not only ban json direct usage of JsonFactory but it also fixes https://github.com/trinodb/trino/issues/17843. Logically these are two different changes it would be nice to keep them as separate commits. Also we should provide test coverage for https://github.com/trinodb/trino/issues/17843

File: client/trino-cli/src/main/java/io/trino/cli/JsonPrinter.java (hunk 8 above)
Comment (martint) [reply client/trino-cli/src/main/java/io/trino/cli/JsonPrinter.java:76]: That's overly broad. Does it suppress *all* modernizer checks?

Comment (wendigo) [reply client/trino-cli/src/main/java/io/trino/cli/JsonPrinter.java:76]: Yes. There is no way to disable a single rule

File: core/trino-main/src/main/java/io/trino/operator/scalar/JsonExtract.java (hunk 12 above)
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/operator/scalar/JsonExtract.java:123]: I'd prefer to make this as a follow up. Canonicalization could and should be disabled in most of the cases when we are working with the external input.

Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/operator/scalar/JsonExtract.java]: Nevermind. I did introduce a constant for that as well.

File: .mvn/modernizer/violations.xml [hunk 14]
```diff
@@ -313,4 +313,16 @@
         <version>1.8</version>
         <comment>Prefer org.testng.annotations.AfterClass</comment>
     </violation>
+
+    <violation>
+        <name>com/fasterxml/jackson/core/JsonFactory."&lt;init&gt;":()V</name>
+        <version>1.8</version>
+        <comment>Use io.trino.plugin.base.util.DefaultJsonFactory.INSTANCE instead</comment>
```
Comment (hashhar) [inline .mvn/modernizer/violations.xml:320]: update these error messages since class names and member names changed.

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java [hunk 15]
```diff
@@ -18,15 +18,23 @@
...
+    public static final JsonFactory NON_CANONICALIZING_JSON_FACTORY = builder()
+            .configure(CANONICALIZE_FIELD_NAMES, false)
+            .configure(INTERN_FIELD_NAMES, false)
+            .build();
+
     private TrinoJsonFactory() {}
 
     @SuppressModernizer
-    public static JsonFactoryBuilder builder()
+    static JsonFactoryBuilder builder()
```
Comment (hashhar) [inline lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java]: should this public to package-private be part of first commit?

**Review by hashhar — APPROVED** [review]

**Review by kokosing — APPROVED** [review]

**Comment by findepi** [comment]
added "Fixes https://github.com/trinodb/trino/issues/17843" in the description

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+    @SuppressModernizer
+    public static JsonFactoryBuilder builder()
+    {
+        return new JsonFactoryBuilder()
+            .streamReadConstraints(StreamReadConstraints.builder()
+                    .maxStringLength(Integer.MAX_VALUE)
+                    .maxNestingDepth(Integer.MAX_VALUE)
+                    .maxNumberLength(Integer.MAX_VALUE)
+                    .build());
+    }
```
Comment (findepi) [inline lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java:45]: should `.configure(JsonGenerator.Feature.AUTO_CLOSE_TARGET, false);` be here?

File: .mvn/modernizer/violations.xml [hunk 17]
```diff
@@ -313,4 +313,16 @@
         <version>1.8</version>
         <comment>Prefer org.testng.annotations.AfterClass</comment>
     </violation>
+
+    <violation>
+        <name>com/fasterxml/jackson/core/JsonFactory."&lt;init&gt;":()V</name>
+        <version>1.8</version>
+        <comment>Use io.trino.plugin.base.util.TrinoJsonFactory.JSON_FACTORY or NON_CANONICALIZING_JSON_FACTORY instead</comment>
```
Comment (findepi) [inline .mvn/modernizer/violations.xml:320]: NON_CANONICALIZING_JSON_FACTORY doesn't exist yet

File: core/trino-main/src/main/java/io/trino/operator/scalar/JsonExtract.java [hunk 18]
```diff
@@ -118,17 +116,13 @@
...
-            .build();
-
     private JsonExtract() {}
 
     public static <T> T extract(Slice jsonInput, JsonExtractor<T> jsonExtractor)
     {
         requireNonNull(jsonInput, "jsonInput is null");
         try {
-            try (JsonParser jsonParser = createJsonParser(JSON_FACTORY, jsonInput)) {
+            try (JsonParser jsonParser = createJsonParser(NON_CANONICALIZING_JSON_FACTORY, jsonInput)) {
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/scalar/JsonExtract.java:125]: it may not be obvious from refactor:

we gained `.configure(INTERN_FIELD_NAMES, false)` here now.

call it out in commit message

**Comment by martint** [comment]
Removed RELEASE-BLOCKER tag since issue is already tagged as such.

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+package io.trino.plugin.base.util;
+
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonFactoryBuilder;
+import com.fasterxml.jackson.core.StreamReadConstraints;
+import org.gaul.modernizer_maven_annotations.SuppressModernizer;
+
+public final class TrinoJsonFactory
+{
+    public static final JsonFactory JSON_FACTORY = builder()
```
[the remark this answers is not in the document]
Comment (chenjian2664) [reply lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java:26]: How about add a function or change to a supplier

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java (hunk 16 above)
Comment (wendigo) [reply lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java:148]: Makes sense. Added

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+import org.gaul.modernizer_maven_annotations.SuppressModernizer;
+
+import static com.fasterxml.jackson.core.StreamWriteFeature.AUTO_CLOSE_TARGET;
+
+public final class TrinoJsonFactory
+        extends JsonFactory
+{
+    private static final String MUTATING_NOT_ALLOWED_MESSAGE = "TrinoJsonFactory does not allow changing configuration. Use TrinoJsonFactory.builder() instead.";
+
+    public static final JsonFactory JSON_FACTORY = new TrinoJsonFactory(builder());
```
Comment (findepi) [inline lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java:36]: How do we know all mutating methods are blocked?
How do we know that, after new jackson version,  all mutating methods are blocked?

What's the benefit of sharing JSON_FACTORY on a public static field? 
At first sight it looks like unneeded complexity. Please explain.

**Review by findepi — CHANGES_REQUESTED** [review]

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java (hunk 20 above)
Comment (findepi) [reply lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java:36]: in other words: maybe remove this field and let everyone use `builder()`.
if someone wants to share json factory (eg for perf reasons), the caller can easily do so on eg a `private [static]` field

Comment (wendigo) [reply lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java:36]: I'll add a test that checks for Jackson version with the comment that if the version has changed - improve the builder/wrapping classes.

Comment (findepi) [reply lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java:35]: please help me find answer to this question:

> What's the benefit of sharing JSON_FACTORY on a public static field?
At first sight it looks like unneeded complexity. Please explain.

**Comment by wendigo** [comment]
@findepi I've dropped most of the changes from this PR and kept bare-minimum.

**Review by chenjian2664 — APPROVED** [review]

**Review by losipiuk — APPROVED** [review]

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+import com.fasterxml.jackson.core.TSFBuilder;
+import com.fasterxml.jackson.core.io.CharacterEscapes;
+import com.fasterxml.jackson.core.io.InputDecorator;
+import com.fasterxml.jackson.core.io.OutputDecorator;
+import org.gaul.modernizer_maven_annotations.SuppressModernizer;
+
+public final class TrinoJsonFactory
+        extends JsonFactory
+{
+    private static final String MUTATING_NOT_ALLOWED_MESSAGE = "TrinoJsonFactory does not allow changing configuration. Use TrinoJsonFactory.builder() instead.";
```
Comment (losipiuk) [inline lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java:29]: What is the benefit of blocking those? 
Feels it should be responsibily of code which created instance of `JsonFactory` to not mutate it. And if the code actually wants to do that - why forbid.

Comment (hashhar) [reply lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java:29]: seems to come as result of https://github.com/trinodb/trino/pull/17854#discussion_r1232227091

Comment (losipiuk) [reply lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java:29]: I do not see how that thread explains why we need to have a subclass of `JsonFactory` which prevents use of mutators. The thread is about droping public shared instance of `TrinoJsonFactory`. While it was needed to have mutators blocked if we had shared instance of `TrinoJsonFactory` managed by the library code, it is no longer useful if we dropped that I think.

**Comment by findepi** [comment]
https://github.com/trinodb/trino/pull/17854#discussion_r1232385848 is resolved, please help me understand what the resolution is. 
What `TrinoJsonFactory` subclass is for?

**Comment by wendigo** [comment]
@findepi disallow mutating already configured JsonFacfory. Instead you should use a builder if you want a different configuration. This makes it easier to reason about factories that are used since they configuration is immutable in the runtime

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java (hunk 21 above)
Comment (findepi) [reply lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java:29]: > We still have shared instances (`JsonUtil` and `JsonOperators`) and I think that this is dangerous to do.

these should become `private static final` fields. maybe something like https://github.com/trinodb/trino/pull/17950 could help

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+import com.fasterxml.jackson.core.JsonFactoryBuilder;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.StreamReadConstraints;
+import com.fasterxml.jackson.core.TSFBuilder;
+import com.fasterxml.jackson.core.io.CharacterEscapes;
+import com.fasterxml.jackson.core.io.InputDecorator;
+import com.fasterxml.jackson.core.io.OutputDecorator;
+import org.gaul.modernizer_maven_annotations.SuppressModernizer;
+
+public final class TrinoJsonFactory
```
Comment (findepi) [inline lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java:26]: https://github.com/trinodb/trino/pull/17950 is merged. please rebase and remove this class

Comment (wendigo) [reply lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java:26]: Rebased but I'm not sure whether allowing JsonFactory mutation is a right thing to do. It's error-prone as it can happen in the runtime, and there is no mechanism that will prohibit cross-class reuse of the existing factories. Ideally I'd delegate that to the airlift and guice to both allow configuring and injecting factories but it's a bigger change as some of those classes are not even DI managed.

**Comment by wendigo** [comment]
@findepi @kokosing dropped the immutability checks. Please review

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+    private TrinoJsonFactory() {}
+
+    @SuppressModernizer
+    public static JsonFactoryBuilder trinoJsonFactoryBuilder()
+    {
+        return trinoJsonFactoryBuilder(new JsonFactory());
+    }
+
+    @SuppressModernizer
+    public static JsonFactoryBuilder trinoJsonFactoryBuilder(JsonFactory factory)
```
Comment (findepi) [inline lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java:32]: inline, it has one usage only

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        return new JsonFactoryBuilder(factory)
+                .streamReadConstraints(StreamReadConstraints.builder()
+                        .maxStringLength(Integer.MAX_VALUE)
+                        .maxNestingDepth(Integer.MAX_VALUE)
+                        .maxNumberLength(Integer.MAX_VALUE)
+                        .build());
+    }
+
+    public static JsonFactory trinoJsonFactory()
```
Comment (findepi) [inline lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java:42]: move to `JsonUtils` in plugin toolkit.

we do not need two different util classes for json, do we?

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        return trinoJsonFactoryBuilder().build();
```
Comment (findepi) [inline lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/TrinoJsonFactory.java:44]: when moving to JsonUtils,  put trinoJsonFactoryBuilder below trinoJsonFactory

File: lib/trino-plugin-toolkit/src/test/java/io/trino/plugin/base/util/TestJsonUtils.java [hunk 26]
```diff
@@ -70,4 +73,24 @@ public void testTrailingContent()
                 .hasMessage("Could not parse JSON")
                 .hasStackTraceContaining("Unrecognized token 'not': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')");
     }
+
+    @Test
+    public void factoryHasNoReadConstraints()
```
Comment (findepi) [inline lib/trino-plugin-toolkit/src/test/java/io/trino/plugin/base/util/TestJsonUtils.java:78]: test methods in this class follow `testFoo` naming convention

File: lib/trino-plugin-toolkit/src/test/java/io/trino/plugin/base/util/TestJsonUtils.java [hunk 27]
```diff
@@ -70,4 +73,24 @@ public void testTrailingContent()
...
+    {
+        assertReadConstraints(jsonFactory().streamReadConstraints());
+        assertReadConstraints(JsonUtils.jsonFactoryBuilder().build().streamReadConstraints());
```
Comment (findepi) [inline lib/trino-plugin-toolkit/src/test/java/io/trino/plugin/base/util/TestJsonUtils.java:81]: When this assertion fails, what should one do?
add code comment explaining why it matters / linking to explanation (eg link to issue)

File: lib/trino-plugin-toolkit/src/test/java/io/trino/plugin/base/util/TestJsonUtils.java [hunk 28]
```diff
@@ -70,4 +73,24 @@ public void testTrailingContent()
...
+
+    @Test
+    public void builderHasNoReadConstraints()
+    {
+        assertReadConstraints(JsonUtils.jsonFactoryBuilder().build().streamReadConstraints());
+    }
+
+    private static void assertReadConstraints(StreamReadConstraints constraints)
+    {
+        AssertionsForClassTypes.assertThat(constraints.getMaxStringLength()).isEqualTo(Integer.MAX_VALUE);
```
Comment (findepi) [inline lib/trino-plugin-toolkit/src/test/java/io/trino/plugin/base/util/TestJsonUtils.java:92]: When this assertion fails, what should one do?
add code comment explaining why it matters / linking to explanation (eg link to issue)

**Review by findepi — COMMENTED** [review]
"Provide safe way to create JsonFactories"

File: .mvn/modernizer/violations.xml [hunk 29]
```diff
@@ -313,4 +313,16 @@
         <version>1.8</version>
         <comment>Prefer org.testng.annotations.AfterClass</comment>
     </violation>
+
+    <violation>
+        <name>com/fasterxml/jackson/core/JsonFactory."&lt;init&gt;":()V</name>
+        <version>1.8</version>
+        <comment>Use io.trino.plugin.base.util.JsonUtils.jsonFactory()</comment>
```
Comment (findepi) [inline .mvn/modernizer/violations.xml:320]: squash "Ban direct usage of JsonFactory and JsonFactoryBuilder" with previous commit

**Comment by findepi** [comment]
lgtm %

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/JsonUtils.java [hunk 30]
```diff
@@ -144,6 +148,22 @@ private static <T> T parseJson(JsonNode node, String jsonPointer, Class<T> javaT
         return jsonTreeToValue(mappingsNode, javaType);
     }
 
+    public static JsonFactory jsonFactory()
+    {
+        return jsonFactoryBuilder().build();
+    }
+
+    @SuppressModernizer
```
[the remark this answers is not in the document]
Comment (findepi) [reply lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/JsonUtils.java:156]: this is resolved, but not applied/answered to. is there another change coming?

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
CI checkstyle failure looks related

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/JsonUtils.java (hunk 30 above)
[the remark this answers is not in the document]
Comment (wendigo) [reply lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/JsonUtils.java:156]: Idk what this comment should contain tbh. JsonPrinter has a comment because it's not reusing JsonUtils.
